torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -4026,134 +4026,43 @@ Tensor linalg_matrix_exp_differential(
-}
-
-// The backward for this function is just a specialized version of
-// lu.backward, which is implemented in /torch/_autograd_functions.py
-Tensor _det_lu_based_helper_backward(
-    const Tensor& det_grad,
+Tensor linalg_det_backward(
+    const Tensor& grad,
     const Tensor& det,
-    const Tensor& self,
-    const Tensor& lu,
-    const Tensor& pivs) {
-  if (!self.numel()) {
-    return at::zeros_like(self, at::MemoryFormat::Contiguous);
-  }
-  if (!det_grad.defined()) {
-    return Tensor();
+    const Tensor& A,
+    const Tensor& LU,
+    const Tensor& pivots) {
+  at::NoTF32Guard disable_tf32;
+  if (!grad.defined()) {
+    return {};
   }
 
-  // run det_backward only if backward is run on _det_lu_based_helper_backward.
-  // _det_lu_based_helper_backward is more stable for forward det computing
-  // functions, but it fails with double backward gradient checks
-  // (gradgradcheck). det_backward, on the other hand, is less stable (due to
-  // restrictions on svd_backward, namely, svd_backward requries distinct
-  // singular values which are sufficiently different from each other), yet, if
-  // its computation is stable, so is its double backward. Hence, if only single
-  // backward is run, we use _det_lu_based_helper_backward, for the double
-  // backward case we use det_backward. The latter approach could produce
-  // unstable gradients, therefore we DO NOT recommend double backpropagation
-  // through det computing functions.
-  if (at::GradMode::is_enabled()) {
-    return det_backward(det_grad, self, det);
+  // The gradient G is the matrix solving
+  // A.mH G = det(A).conj() * grad * I
+  auto d_diag = grad * det.conj();
+  auto d = at::diag_embed(d_diag.unsqueeze(-1).expand(pivots.sizes()));
+
+  if (!at::GradMode::is_enabled()) {
+    // The formula is given by the solution of AX = det.conj() * det * I when A
+    // is invertible det is C^1, so if it's not invertible, we can wiggle the LU
+    // decomposition a bit and use the resulting matrix as a decent
+    // approximation
+    auto eps = at::native::_get_epsilon(c10::toRealValueType(LU.scalar_type()));
+    auto LU_ =
+        LU + at::diag_embed(at::where(LU.diagonal(0, -2, -1) == 0., eps, 0.));
+    return at::linalg_lu_solve(LU_, pivots, d, /*left=*/true, /*adjoint=*/true);
+  } else {
+    // TODO In this case, the trick above just does not cut it
+    // The proper way of doing this is doing `auto mask = det == 0.;` and then
+    // if any determinant is zero, use an SVD decomposition to compute the
+    // derivative in those inputs (not all inputs). The derivative may be then
+    // computed explicitly by noting that the gradient of the derivative of the
+    // determinant is given by adj(A^H)* grad where adj(A) is the adjugate of A.
+    // Then, the adjugate of a singular matrix may be computed as per
+    // https://nhigham.com/2020/06/16/what-is-the-adjugate-of-a-matrix/
+    // adj(A) = conj(det(V)) * det(U) * (S[:-1].prod()) * U[-1] @ Vh[-1]
```

Comment:
`S[1:]`?

Reply:
`S[:-1]`, as we want to remove the smallest singular value, that is, the last one.

Reply:
Right, sorry, forgot the order :)